keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -118,78 +142,79 @@ Currently, kubeadm vendors in pieces of Go code from the components to be able t
 If a target component does not have any means to verify its config, then kubeadm won’t do any validation for the time being. It will instead display a warning to the user denoting that certain component’s config is unvalidated.
 
-### Kubernetes Core vs AddOns Component Configs
+### User supplied vs kubeadm generated component configs
 
-Currently kubeadm maintains the component configs of both kubelet and kube-proxy. These are components that are maintained in very different ways.
+Component configs can be specified by users in their entirety along with the kubeadm config during init or in a config map. Hewever, most users choose not to supply the configs themselves, but to trust kubeadm to generate correct configs. There are also users who would let kubeadm generate the configs and then edit them in place in the config maps that house them.
 
-Kubelet, along with API Server, Scheduler, Controller Manager and etcd, are deemed essential Kubernetes components. They are deployed and managed locally by kubeadm, using systemd services (kubelet) or static pods on the local kubelet (API Server, Scheduler, etc.). Hence, their configs are usually consumed by the components in the form of local files.
+Because kubeadm looses its ability to automatically convert an older component config version into a newer one, it is best to start distinguishing between user supplied or user modified component configs and kubeadm generated ones. That way kubeadm generated configs can be thrown out and regenerated when an upgrade is required.
 
-On the other hand, kube-proxy and CoreDNS/kube-dns are addons, that are managed as standard Kubernetes deployments. Hence their configs are usually accessed via config maps.
+Unfortunately, user supplied or user modified configs would still need to be manually migrated if a config version upgrade is necessary.
 
-Therefore, two different config lifecycles are distinguishable here - one for Kubernetes core components and one for cluster addons.
+To distinguish between user supplied and generated configs, kubeadm shall sign each config map, that contains generated by it config, during cluster upload. If the signature is missing or incorrect during download the config is treated as user supplied.
 
-The defaulting and validation changes, proposed in the previous paragraphs, are going to be applied to both Kubernetes core and AddOn component config types.
+To calculate the signature of a generated config, kubeadm would prepare the standard `core/v1` config map. Then it would take the `data` section, order the keys lexicographically and then feed the values in that order to a SHA256 digest. After that, the same procedure is repeated for the values in the `binaryData` section too.
 
-#### AddOn Component Config management
+The resulting SHA256 digest is prepended its type - the string `sha256:` and is stored as a value of the `kubeadm.kubernetes.io/autogen-checksum` annotation in the config map object.
```

Comment:
nit
```suggestion
The resulting SHA256 digest is prepended its type - the string `sha256:` and is stored as a value of the `kubeadm.kubernetes.io/hash` annotation in the config map object.
```
for consistency with `kubeadm.controlplane.cluster.x-k8s.io/hash` (this can be fixed also during implementation)

Reply:
It was probably a mistake on my part to put it in here in this form. As you said, it's really a nit and this KEP can be retroactively updated with the exact nits once we merge the PR that brings this functionality to kubeadm (kubernetes/kubernetes#86070)